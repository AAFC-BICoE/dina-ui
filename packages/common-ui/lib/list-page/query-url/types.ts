export interface SimpleQueryGroup {
  // Conjunction (AND/OR)
  c: string;

  // Props: Items inside the group
  p: SimpleQueryRow[];
}

export interface SimpleQueryRow {
  // Field
  f: string;

  // Operator
  o: string;

  // Value
  v: string;

  // Type
  t: string;

  // Dynamic field identifier (UUID for example if managed attribute)
  d?: string;
}
