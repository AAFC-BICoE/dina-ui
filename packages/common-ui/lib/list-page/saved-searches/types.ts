import { JsonTree } from "react-awesome-query-builder";

/**
 * Since saved searches are stored as JSON this is the structure to be used.
 */
export interface SavedSearchStructure {
  [indexName: string]: {
    [savedSearchName: string]: SingleSavedSearch;
  };
}

export interface SingleSavedSearch {
  // This field is not saved in the JSON.
  savedSearchName?: string;
  default: boolean;
  queryTree?: JsonTree;
}
