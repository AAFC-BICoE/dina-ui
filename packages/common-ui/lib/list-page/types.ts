import { KitsuResource } from "kitsu";
import { Column } from "react-table";

/**
 * This type extends the react-table column type, this just adds a few specific fields for elastic
 * search mapping and internationalization.
 */
export interface TableColumn<TData extends KitsuResource>
  extends Column<TData> {
  /**
   * User-friendly column to be displayed. You can use a DinaMessage key for internationalization.
   */
  label?: string;

  /**
   * Elastic search path to the attribute.
   *
   * Example: `data.attributes.name`
   */
  attributePath?: string;

  /**
   * This field is used to find the relationship in the included section.
   */
  relationshipType?: string;

  /**
   * Is this attribute considered a keyword in elastic search. Required for filtering and sorting.
   */
  isKeyword?: boolean;

  /**
   * The QueryPage will only display the accessors that are displayed on the result table. However,
   * if you have custom cells that receive other fields you will need to add them to this list so
   * elastic search includes the fields in the result.
   *
   * Example: `data.attributes.name`
   *
   * Please note that duplicate fields are automatically removed so you don't need to worry about
   * having unique accessors.
   */
  additionalAccessors?: string[];
}

/**
 * The full path will be generated for elastic using a combination of the parent path and
 * the value. The path is generated using the following:
 *
 * {parentPath}.{path}.{value}
 *
 * Example: included.attributes.determination.verbatimScientificName
 */
export interface ESIndexMapping {
  /**
   * Name of the attribute.
   *
   * Example: verbatimScientificName
   */
  value: string;

  /**
   * Text that is displayed to the user in the Query Filtering option menu.
   *
   * This text is a user-friendly generated label, which may show some paths to help the user
   * understand the relationships better. This is generated from the path.
   *
   * Example: determination.verbatimScientificName
   */
  label: string;

  /**
   * The attributes type. This can change how the query row is displayed and the options provided.
   *
   * Examples: text, keyword, boolean, date, boolean, long, short, integer...
   */
  type: string;

  /**
   * If enabled, it will allow the user to see suggestions as they type. The suggestions will come
   * from elastic search based on most common values saved.
   *
   * Only available for the text type.
   */
  distinctTerm: boolean;

  /**
   * If enabled it will allow the user to search based on the starting of a word.
   *
   * Example: Hexapoda can be matched with "hex".
   */
  startsWithSupport: boolean;

  /**
   * If enabled it will allow the user to search based in the middle of a word.
   *
   * Example: Hexapoda can be matched with "pod".
   */
  containsSupport: boolean;

  /**
   * If enabled it will allow the user to search based on the ending of a word.
   *
   * Example: Hexapoda can be matched with "poda".
   */
  endsWithSupport: boolean;

  /**
   * The path for the attribute without the attribute name. This path does not include the parent
   * path.
   *
   * Example: attribute.determination
   */
  path: string;

  /**
   * If the attribute belongs to a relationship, this is the path for only the parent. When generating
   * the elastic search query it will use this as the prefix of the path.
   *
   * Example: included
   */
  parentPath?: string;

  /**
   * If the attribute belongs to a relationship, this is the name which will be used to group
   * attributes under the same relationship together in the search. This name will also be used to
   * display text of the group.
   *
   * This text should match the user-friendly label in the locales files. It will be searched using
   * title_[parentName].
   *
   * Example: preparationMethod (and the label will be `title_preparationMethod`)
   */
  parentName?: string;

  /**
   * The parent type is the relationship type to be used. This will be used for elastic search
   * filtering. The reason this is used since the same field can be used for multiple relationships.
   *
   * Example: preparation-method
   */
  parentType?: string;

  /**
   * Only provided if it was added using a dynamic field config.
   */
  dynamicField?: DynamicField;
}

/**
 * Each type on the query builder has a function that is used to transform the query builder row
 * into elastic search logic.
 *
 * @see QueryBuilderElasticSearchExport
 * @see QueryRowTextSearch
 */
export interface TransformToDSLProps {
  /**
   * The query builder type, not really used but required in the function for the QueryBuilder
   * library.
   *
   * For the Elastic Search type you can use the fieldInfo.type instead.
   */
  queryType: string;

  /**
   * The value to search against. If the type is "text" then this is the text that will be filtered.
   */
  value: string;

  /**
   * The operation being performed.
   *
   * For example:
   * "equals", "notEquals", "containsText"
   *
   * Operators are defined in the QueryBuilderConfig file.
   */
  operation: string;

  /**
   * This field path is the unique path given to each item in the field list.
   *
   * For getting the proper absolute field path, use the FieldInfo to generate it.
   */
  fieldPath: string;

  /**
   * The elastic search mapping for the field.
   */
  fieldInfo?: ESIndexMapping;
}

export type DynamicFieldType = "managedAttribute" | "fieldExtension";

export interface DynamicFieldsMappingConfig {
  /** Attribute level dynamic fields */
  fields: DynamicField[];

  /** Dynamic fields for relationships */
  relationshipFields: RelationshipDynamicField[];
}

export interface DynamicField {
  /**
   * Option label that should be used.
   */
  label: string;

  type: DynamicFieldType;

  path: string;

  /**
   * Endpoint where these dynamic fields can be retrieved to list.
   *
   * Example: "collection-api/managed-attribute"
   */
  apiEndpoint: string;

  /**
   * Optional field to indicate which Managed Attributes or Field Extensions should be listed.
   */
  component?: string;
}

/**
 * Configuration for where the Dynamic Field can be found within the relationship index mapping.
 */
export interface RelationshipDynamicField extends DynamicField {
  referencedBy: string;
  referencedType: string;
}
