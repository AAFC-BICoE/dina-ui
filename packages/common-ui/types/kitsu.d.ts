/// <reference types="node" />

import { AxiosInstance } from "axios";
import { SetRequired } from "type-fest";

declare module "kitsu" {
  // export default Kitsu;

  /** JSONAPI client. */
  export default class Kitsu {
    public axios: AxiosInstance;

    public headers: any;

    constructor(params: KitsuConstructorParams);

    public delete(...args: any[]): Promise<any>;

    public get<TData, TMeta = undefined>(
      path: string,
      params: GetParams
    ): Promise<KitsuResponse<TData, TMeta>>;

    public patch(...args: any[]): Promise<any>;

    public post(...args: any[]): Promise<any>;
  }

  /** Params for the Kitsu JSONAPI client's constructor */
  export interface KitsuConstructorParams {
    baseURL: string;
    headers?: any;
    pluralize?: boolean;
    camelCaseTypes?: boolean;
    resourceCase?: "kebab" | "snake" | "none";
  }

  /** Parameters for GET requests. */
  export interface GetParams {
    /** Fields to include in the response data. */
    fields?: FieldsParam;

    /** Resource filter */
    filter?: FilterParam;

    /**
     * Sort order + attribute.
     * Examples:
     *  - name
     *  - -description
     */
    sort?: string;

    /** Included resources. */
    include?: string;

    /** Vendor-specific parameter for paginating listed data. */
    page?: any;
  }

  /** Parameter for requesting sparse fields. */
  export interface FieldsParam {
    [key: string]: string;
  }

  /** Parameter for filtering listed data. */
  export type FilterParam = string | Record<string, string>;

  /** The response from a Kitsu GET request. */
  export interface KitsuResponse<
    TData extends KitsuResponseData,
    TMeta = undefined
  > {
    data: TData extends (infer R)[]
      ? PersistedResource<R>[]
      : PersistedResource<TData>;
    meta: TMeta;
  }

  /** The Kitsu response data can either be one resource or an array of resources. */
  export type KitsuResponseData = KitsuResource | KitsuResource[];

  /** JSONAPI resource base attributes. */
  export interface KitsuResource {
    id?: string;
    type: string;
  }

  export interface KitsuResourceLink {
    id: string | null;
    type: string;
  }

  /**
   * Makes the 'id' field required on a resource type and all of its relationships.
   * Used when assuming that data from the back-end always has the ID set.
   */
  export type PersistedResource<TData extends KitsuResource = KitsuResource> = {
    [P in keyof TData]: TData[P] extends KitsuResource
      ? PersistedResource<TData[P]>
      : TData[P] extends KitsuResource | undefined
      ? PersistedResource<TData[P]> | undefined
      : TData[P];
  } &
    Required<KitsuResource>;

  /**
   * Used when creating or updating a resource.
   * Makes the 'id' field optional on the main resource.
   * Makes the 'id' field required on linked resources.
   * 'type' must be defined.
   */
  export type InputResource<TData extends KitsuResource> = SetRequired<
    {
      [P in keyof TData]?: TData[P] extends KitsuResource | undefined
        ? Required<KitsuResourceLink> | undefined
        : TData[P];
    },
    "type"
  >;
}
