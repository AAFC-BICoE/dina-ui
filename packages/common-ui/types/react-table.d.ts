import "@tanstack/react-table";

declare module "@tanstack/table-core" {
  interface ColumnMeta<TData extends RowData, TValue> {
    className?: string;
    style?: CSSProperties;
  }

  interface TableMeta<TData extends RowData> {
    updateData?: (rowIndex: number, columnId: string, value: unknown) => void;
    enableEditing?: boolean;
  }
}
