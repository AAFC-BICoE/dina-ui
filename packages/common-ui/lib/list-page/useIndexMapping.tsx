import { useEffect, useState } from "react";
import { useApiClient } from "..";
import { DynamicFieldsMappingConfig, ESIndexMapping } from "./types";

export interface UseIndexMappingProps {
  indexName: string;

  /**
   * This is used to indicate to the QueryBuilder all the possible places for dynamic fields to
   * be searched against. It will also define the path and data component if required.
   *
   * Dynamic fields are like Managed Attributes or Field Extensions where they are provided by users
   * or grouped terms.
   */
  dynamicFieldMapping?: DynamicFieldsMappingConfig;
}

/**
 * Custom hook for retrieving the index mapping.
 *
 * The index mapping is used to get all of the fields supported for searching. Also contains
 * additional information on each field.
 *
 * @param indexName The index to retrieve. Example: `dina-material-sample-index`
 */
export function useIndexMapping({
  indexName,
  dynamicFieldMapping
}: UseIndexMappingProps) {
  const { apiClient } = useApiClient();

  // State to store the index map after it has been retrieved.
  const [indexMap, setIndexMap] = useState<ESIndexMapping[]>();

  // Retrieve the index mapping on the first hook load.
  useEffect(() => {
    async function getIndexMapping() {
      const mapping = await fetchQueryFieldsByIndex();
      setIndexMap(mapping);
    }
    getIndexMapping();
  }, []);

  /**
   * The query builder options are generated from the elastic search index. This method will
   * request the mappings from the index.
   *
   * @param searchIndexName index to retrieve from.
   * @returns ESIndexMapping[]
   */
  async function fetchQueryFieldsByIndex() {
    try {
      const resp = await apiClient.axios.get("search-api/search-ws/mapping", {
        params: { indexName }
      });

      // Fields that are dynamic do not need to be listed here.
      const fieldsToSkip =
        dynamicFieldMapping?.fields?.map<string>(
          (fieldMapping) => fieldMapping.path
        ) ?? [];

      const result: ESIndexMapping[] = [];

      // Read index attributes.
      resp.data?.attributes
        ?.filter((key) => key.name !== "type")
        .map((key) => {
          const path = key.path;
          const prefix = "data.attributes";
          let attrPrefix;
          if (path && path.includes(prefix)) {
            attrPrefix = path.substring(prefix.length + 1);
          }

          // Manually remove managed attributes and extension fields from here,
          // they are handled using the dynamic mapping config. See the dynamicFieldMapping.
          if (!fieldsToSkip.some((skipPath) => path.startsWith(skipPath))) {
            result.push({
              label: attrPrefix ? attrPrefix + "." + key.name : key.name,
              value: key.path
                ? key.path + "." + key.name
                : key.name === "id"
                ? "data." + key.name
                : key.name,
              type: key.type,
              path: key.path,

              // Additional options for the field:
              distinctTerm: key.distinct_term_agg,
              startsWithSupport: key?.fields?.includes("prefix") ?? false,
              containsSupport: key?.fields?.includes("infix") ?? false,
              endsWithSupport: key?.fields?.includes("prefix_reverse") ?? false
            });
          }
        });

      // Read relationship attributes.
      resp.data?.relationships?.map((relationship) => {
        relationship?.attributes?.map((relationshipAttribute) => {
          // This is the user-friendly label to display on the search dropdown.
          const attributeLabel = relationshipAttribute.path?.includes(".")
            ? relationshipAttribute.path.substring(
                relationshipAttribute.path.indexOf(".") + 1
              ) +
              "." +
              relationshipAttribute.name
            : relationshipAttribute.name;

          result.push({
            label: attributeLabel,
            value: relationship.value + "." + attributeLabel,
            type: relationshipAttribute.type,
            path: relationshipAttribute.path,
            parentName: relationship.referencedBy,
            parentType: relationship.value,
            parentPath: relationship.path,

            // Additional options for the field:
            distinctTerm: relationshipAttribute.distinct_term_agg,
            startsWithSupport:
              relationshipAttribute?.fields?.includes("prefix") ?? false,
            containsSupport:
              relationshipAttribute?.fields?.includes("infix") ?? false,
            endsWithSupport:
              relationshipAttribute?.fields?.includes("prefix_reverse") ?? false
          });
        });
      });

      // Inject dynamic field mapping config into these.
      if (dynamicFieldMapping) {
        dynamicFieldMapping.fields.forEach((fieldMapping) => {
          result.push({
            dynamicField: fieldMapping,
            value: fieldMapping.path,
            distinctTerm: false,
            label: fieldMapping.label,
            path: fieldMapping.path,
            type: fieldMapping.type,
            startsWithSupport: false,
            containsSupport: false,
            endsWithSupport: false
          });
        });
        dynamicFieldMapping.relationshipFields.forEach(
          (relationshipFieldMapping) => {
            result.push({
              dynamicField: relationshipFieldMapping,
              parentName: relationshipFieldMapping.referencedBy,
              parentPath: "included",
              parentType: relationshipFieldMapping.referencedType,
              value:
                relationshipFieldMapping.path +
                "_" +
                relationshipFieldMapping.referencedBy,
              distinctTerm: false,
              label: relationshipFieldMapping.label,
              path: relationshipFieldMapping.path,
              type: relationshipFieldMapping.type,
              startsWithSupport: false,
              containsSupport: false,
              endsWithSupport: false
            });
          }
        );
      }

      return result;
    } catch (error) {
      return undefined;
    }
  }

  return { indexMap };
}
