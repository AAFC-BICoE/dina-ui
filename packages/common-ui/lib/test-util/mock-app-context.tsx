import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { AccountContextI, AccountProvider } from "../account/AccountProvider";
import {
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider
} from "../api-client/ApiClientContext";
import { CommonUIIntlProvider } from "../intl/common-ui-intl";
import { ModalProvider } from "../modal/modal";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <CommonUIIntlProvider>
            <div ref={modalWrapperRef}>
              <ModalProvider appElement={modalWrapperRef.current}>
                {children}
              </ModalProvider>
            </div>
          </CommonUIIntlProvider>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
