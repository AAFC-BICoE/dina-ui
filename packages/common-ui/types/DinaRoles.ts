export const DINA_ADMIN = "dina-admin";
export const READ_ONLY_ADMIN = "read-only-admin";
export const SUPER_USER = "super-user";
export const USER = "user";
export const GUEST = "guest";
export const READ_ONLY = "read-only";
