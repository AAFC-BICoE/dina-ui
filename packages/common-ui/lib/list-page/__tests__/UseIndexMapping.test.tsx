import { useEffect } from "react";
import { mountWithAppContext } from "../../test-util/mock-app-context";
import { ESIndexMapping } from "../types";
import { useIndexMapping } from "../useIndexMapping";

const INDEX_NAME = "dina-material-sample-index";

interface UseIndexMappingWrapperProps {
  // Callback when the index mapping is retrieved.
  indexMapRetrieved: (indexMap: ESIndexMapping[]) => void;
}

/**
 * Since we are testing a react hook, we will need to create a component that we can use
 * to retrieve the data from the hook.
 *
 * @param indexMapRetrieved Callback when the index mapping is retrieved.
 * @returns Blank component.
 */
function UseIndexMappingWrapper({
  indexMapRetrieved
}: UseIndexMappingWrapperProps) {
  const { indexMap } = useIndexMapping({
    indexName: INDEX_NAME
  });

  useEffect(() => {
    if (indexMap) {
      indexMapRetrieved(indexMap);
    }
  }, [indexMap]);

  return <></>;
}

const mockIndexMapRetrieved = jest.fn();

const mockSearchApiGet = jest.fn<any, any>((path) => {
  if (path === "search-api/search-ws/mapping") {
    return {
      data: {
        attributes: [
          {
            name: "materialSampleType",
            type: "text",
            path: "data.attributes",
            distinct_term_agg: true
          },
          { name: "createdOn", type: "date", path: "data.attributes" },
          {
            name: "materialSampleName",
            type: "text",
            path: "data.attributes",
            fields: ["prefix", "infix", "prefix_reverse"]
          },
          { name: "preparationDate", type: "date", path: "data.attributes" },
          { name: "tags", type: "text", path: "data.attributes" },
          { name: "createdBy", type: "text", path: "data.attributes" }
        ],
        relationships: [
          {
            referencedBy: "collectingEvent",
            name: "type",
            path: "included",
            value: "collecting-event",
            attributes: [
              { name: "createdBy", type: "text", path: "attributes" },
              { name: "createdOn", type: "date", path: "attributes" },
              { name: "tags", type: "text", path: "attributes" },
              { name: "habitat", type: "text", path: "attributes" },
              { name: "substrate", type: "text", path: "attributes" },
              {
                name: "dwcOtherRecordNumbers",
                type: "text",
                path: "attributes"
              },
              { name: "dwcRecordNumber", type: "text", path: "attributes" },
              { name: "startEventDateTime", type: "date", path: "attributes" },
              { name: "endEventDateTime", type: "date", path: "attributes" }
            ]
          },
          {
            referencedBy: "preparationMethod",
            name: "type",
            path: "included",
            value: "preparation-method",
            attributes: [
              {
                name: "name",
                type: "text",
                path: "attributes",
                distinct_term_agg: true
              }
            ]
          },
          {
            referencedBy: "storageUnit",
            name: "type",
            path: "included",
            value: "storage-unit",
            attributes: [{ name: "name", type: "text", path: "attributes" }]
          },
          {
            referencedBy: "organism",
            name: "type",
            path: "included",
            value: "organism",
            attributes: [
              {
                name: "verbatimScientificName",
                type: "text",
                path: "attributes.determination"
              },
              {
                name: "scientificName",
                type: "text",
                path: "attributes.determination"
              },
              {
                name: "typeStatus",
                type: "text",
                path: "attributes.determination"
              }
            ]
          },
          {
            referencedBy: "collection",
            name: "type",
            path: "included",
            value: "collection",
            attributes: [
              {
                name: "name",
                type: "text",
                path: "attributes",
                distinct_term_agg: true
              },
              {
                name: "code",
                type: "text",
                path: "attributes",
                distinct_term_agg: true
              }
            ]
          },
          {
            referencedBy: "assemblages",
            name: "type",
            path: "included",
            value: "assemblage",
            attributes: [{ name: "name", type: "text", path: "attributes" }]
          },
          {
            referencedBy: "preparationType",
            name: "type",
            path: "included",
            value: "preparation-type",
            attributes: [
              {
                name: "name",
                type: "text",
                path: "attributes",
                distinct_term_agg: true
              }
            ]
          }
        ],
        index_name: "dina_material_sample_index"
      }
    };
  }
});

describe("Use Index Mapping Hook", () => {
  it("Retrieve index and transform the structure.", async () => {
    const wrapper = mountWithAppContext(
      <UseIndexMappingWrapper
        indexMapRetrieved={(newIndexMap: any) => {
          mockIndexMapRetrieved(newIndexMap);
        }}
      />,
      {
        apiContext: {
          apiClient: {
            axios: { get: mockSearchApiGet } as any
          }
        }
      }
    );

    await new Promise(setImmediate);
    wrapper.update();

    expect(mockIndexMapRetrieved).toMatchSnapshot();
  });
});
