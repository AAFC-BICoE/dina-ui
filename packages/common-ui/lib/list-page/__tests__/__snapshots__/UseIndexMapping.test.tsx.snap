// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Use Index Mapping Hook Retrieve index and transform the structure. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "containsSupport": false,
          "distinctTerm": true,
          "endsWithSupport": false,
          "label": "materialSampleType",
          "path": "data.attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "data.attributes.materialSampleType",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "createdOn",
          "path": "data.attributes",
          "startsWithSupport": false,
          "type": "date",
          "value": "data.attributes.createdOn",
        },
        Object {
          "containsSupport": true,
          "distinctTerm": undefined,
          "endsWithSupport": true,
          "label": "materialSampleName",
          "path": "data.attributes",
          "startsWithSupport": true,
          "type": "text",
          "value": "data.attributes.materialSampleName",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "preparationDate",
          "path": "data.attributes",
          "startsWithSupport": false,
          "type": "date",
          "value": "data.attributes.preparationDate",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "tags",
          "path": "data.attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "data.attributes.tags",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "createdBy",
          "path": "data.attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "data.attributes.createdBy",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "createdBy",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "collecting-event.createdBy",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "createdOn",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "date",
          "value": "collecting-event.createdOn",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "tags",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "collecting-event.tags",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "habitat",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "collecting-event.habitat",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "substrate",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "collecting-event.substrate",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "dwcOtherRecordNumbers",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "collecting-event.dwcOtherRecordNumbers",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "dwcRecordNumber",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "collecting-event.dwcRecordNumber",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "startEventDateTime",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "date",
          "value": "collecting-event.startEventDateTime",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "endEventDateTime",
          "parentName": "collectingEvent",
          "parentPath": "included",
          "parentType": "collecting-event",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "date",
          "value": "collecting-event.endEventDateTime",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": true,
          "endsWithSupport": false,
          "label": "name",
          "parentName": "preparationMethod",
          "parentPath": "included",
          "parentType": "preparation-method",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "preparation-method.name",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "name",
          "parentName": "storageUnit",
          "parentPath": "included",
          "parentType": "storage-unit",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "storage-unit.name",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "determination.verbatimScientificName",
          "parentName": "organism",
          "parentPath": "included",
          "parentType": "organism",
          "path": "attributes.determination",
          "startsWithSupport": false,
          "type": "text",
          "value": "organism.determination.verbatimScientificName",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "determination.scientificName",
          "parentName": "organism",
          "parentPath": "included",
          "parentType": "organism",
          "path": "attributes.determination",
          "startsWithSupport": false,
          "type": "text",
          "value": "organism.determination.scientificName",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "determination.typeStatus",
          "parentName": "organism",
          "parentPath": "included",
          "parentType": "organism",
          "path": "attributes.determination",
          "startsWithSupport": false,
          "type": "text",
          "value": "organism.determination.typeStatus",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": true,
          "endsWithSupport": false,
          "label": "name",
          "parentName": "collection",
          "parentPath": "included",
          "parentType": "collection",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "collection.name",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": true,
          "endsWithSupport": false,
          "label": "code",
          "parentName": "collection",
          "parentPath": "included",
          "parentType": "collection",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "collection.code",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": undefined,
          "endsWithSupport": false,
          "label": "name",
          "parentName": "assemblages",
          "parentPath": "included",
          "parentType": "assemblage",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "assemblage.name",
        },
        Object {
          "containsSupport": false,
          "distinctTerm": true,
          "endsWithSupport": false,
          "label": "name",
          "parentName": "preparationType",
          "parentPath": "included",
          "parentType": "preparation-type",
          "path": "attributes",
          "startsWithSupport": false,
          "type": "text",
          "value": "preparation-type.name",
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
